File: bench/splatexr/splatexr.cpp
#include <vector>
#include <OpenEXR/ImfRgbaFile.h>
#include <functional>

using namespace std;

int write_rgba(int width, 
    int height, 
    const char* filename, 
    const float* raw_rgbas) {

  std::vector<Imf::Rgba> pixels(width * height);
  for (int p = 0; p < width * height; ++p) {
    pixels[p] = Imf::Rgba(raw_rgbas[4*p+0], 
                          raw_rgbas[4*p+1], 
                          raw_rgbas[4*p+2],
                          raw_rgbas[4*p+3]);
  }
  Imf::RgbaOutputFile file (filename, width, height, Imf::WRITE_RGBA);
  file.setFrameBuffer(&pixels[0], 1, width);
  file.writePixels (height); 
  return 1;
}

struct Pixel {
  float _xyz[3];
  float _weight;
};

struct Film {
  std::vector<Pixel> _pixels;

  Film(int w, int h):
    _w(w),
    _h(h),
    _pixels(w * h)
  {}

  void set_pixel(Pixel p, int x, int y){
    _pixels[y * _w + x] = p;
  }

  int _w;
  int _h;
};

void image_process(Film& film, std::function<Pixel(int,int)> f) {

  for (int y = 0; y < film._h; ++y){
    for (int x = 0; x < film._w; ++x) {
      film.set_pixel(f(x,y), x, y);
    }
  }

}

Pixel boxes_fn(int x, int y) {
  return {{(float)sin(0.01 * 0.01 * x * y), 0, 0}, 1};
}

Pixel bench_fn(int x, int y) {
  return {{(float)sin(0.03 * 0.03 * x * y), 0, 0}, 1};
}
int main(int argc, char** argv) {

  int w = 512;
  int h = 512;
  int iterations = 10;
  if (argc > 1) {
    iterations = atoi(argv[1]);
  }

  printf("%d iterations\n", iterations);
  Film film(w, h);
  for (int i = 0; i < iterations; ++i) {
    image_process(film, bench_fn);
  }
  write_rgba(w, h, "/tmp/cpp.exr", (const float*)&film._pixels[0]);
  return 0;
}

File: bench/jsplatexr/SplatExr.java
import com.sun.jna.Library;
import com.sun.jna.Native;
import com.sun.jna.Platform;

/** Simple example of JNA interface mapping and usage. */
public class SplatExr {

  // This is the standard, stable way of mapping, which supports extensive
  // customization and mapping of Java to native types.
  public interface CLibrary extends Library {
    CLibrary INSTANCE = (CLibrary)
      Native.loadLibrary("exru", CLibrary.class);

    void write_rgba(int width, int height, String filename, float[] raw_rgbas);
  }

  public static void image_process(Film film) {

  
    for (int y = 0; y < film._height; ++y){
      for (int x = 0; x < film._width; ++x) {

        Pixel p = new Pixel((float)(Math.sin(0.1 * x) * Math.sin(0.1 * y)),
                            0.0f, 
                            0.0f, 
                            1.0f);
        film.set_pixel(p, x, y);
      }
    }
  }

  public static void main(String[] args) {
    // System.setProperty("jna.Library.path", "/some/absolute/path");

    int w = 512;
    int h = 512;

    int iterations = args.length < 1 ? 10 : Integer.parseInt(args[0]);
    System.out.println("Iterations " + iterations);
    for (int n = 0; n < iterations; ++n) {
      Film f = new Film(w, h);
      image_process(f);
    }

      Film f = new Film(w, h);
      image_process(f);
    System.out.println("Writing to /tmp/jsplat.exr...");
    CLibrary.INSTANCE.write_rgba(w,h, "/tmp/jsplat.exr", f.write_floats());
  }
}

